official/utils/logging/logger.py

```diff
@@ -72,3 +84,106 @@ def log_metric(self, name, value, unit=None, global_step=None, extras=None):
       except (TypeError, ValueError) as e:
         tf.logging.warning("Failed to dump metric to log file: "
                            "name %s, value %s, error %s", name, value, e)
+
+  def log_run_info(self, model_name):
+    """Collect most of the TF runtime information for the local env.
+
+    The schema of the run info follows official/benchmark/datastore/schema.
+    """
+    run_info = {"model_name": model_name}
+    _collect_tensorflow_info(run_info)
+    _collect_environment_variable(run_info)
+    _collect_cpu_info(run_info)
+    _collect_gpu_info(run_info)
+    _collect_memory_info(run_info)
+
+    with tf.gfile.GFile(os.path.join(
+        self._logging_dir, _BENCHMARK_RUN_LOG_FILE_NAME), "w") as f:
+      try:
+        json.dump(run_info, f)
+        f.write("\n")
+      except (TypeError, ValueError) as e:
+        tf.logging.warning("Failed to dump benchmark run info to log file: %s",
+                           e)
+
+def _collect_tensorflow_info(run_info):
+  run_info["tensorflow_version"] = {
+      "version": tf.VERSION, "git_hash": tf.GIT_VERSION}
+
+def _collect_environment_variable(run_info):
+  run_info["environment_variable"] = {
+      k:v for k, v in os.environ.items() if k.startswith("TF_")}
+
+# The following code is mirrored from tensorflow/tools/test/system_info_lib
+# which is not exposed for import.
+def _collect_cpu_info(run_info):
+  cpu_info = {}
+
+  cpu_info["num_cores"] = multiprocessing.cpu_count()
+  # Gather num_cores_allowed
+  try:
+    with tf.gfile.GFile("/proc/self/status", "rb") as fh:
+      nc = re.search(r"(?m)^Cpus_allowed:\s*(.*)$", fh.read())
```

Comment:
Oof. This is scary. What's happening here? Is there a better way? What exactly is this regex doing, and is it dangerous to read in an arbitrary file? Does this limit us to particular OSes?

Reply:
It is trying to read the local os state to get CPU allowed, You can find the format and content from https://www.kernel.org/doc/Documentation/filesystems/proc.txt and it is not an arbitrary file. I think it works on Linux/Mac system, and Windows will probably fail.

The regex is trying to din the line with Cpus_allowed as start and find the matching group in the bracket.

Reply:
What about something like this instead? https://stackoverflow.com/questions/1006289/how-to-find-out-the-number-of-cpus-using-python ? Or https://docs.python.org/3/library/platform.html#cross-platform . I am inclined to think that the set of information we want should be limited to the set accessible from the default Python libs, rather than adding extra dependencies or very complex regexes to get platform-specific info that we're not sure we need yet. Especially since we will control the hardware we are running on. Can we start conservative, and then add the additional complexity if we find we need it down the line?

Reply:
Done. Removed most of the regex parse to eliminate potential error.